packages/trackerless-network/src/logic/proxy/ProxyStreamConnectionClient.ts

```diff
@@ -0,0 +1,257 @@
+        const proxiesToAttempt = sampleSize(Array.from(this.definition!.peers.keys()).filter((id) =>
+            !this.connections.has(id)
+        ), connectionCount)
+        await Promise.all(proxiesToAttempt.map((id) =>
+            this.attemptConnection(id, this.definition!.direction, this.definition!.userId)
+        ))
+    }
+
+    private async attemptConnection(peer: PeerIDKey, direction: ProxyDirection, userId: UserID): Promise<void> {
+        const peerDescriptor = this.definition!.peers.get(peer)!
+        const client = toProtoRpcClient(new ProxyConnectionRpcClient(this.rpcCommunicator.getRpcClientTransport()))
+        const proxyPeer = new RemoteProxyServer(peerDescriptor, this.config.streamPartId, client)
+        const accepted = await proxyPeer.requestConnection(this.config.ownPeerDescriptor, direction, userId)
+        if (accepted) {
+            this.config.connectionLocker.lockConnection(peerDescriptor, 'proxy-stream-connection-client')
+            this.connections.set(peer, direction)
+            const remote = new RemoteRandomGraphNode(
+                peerDescriptor,
+                this.config.streamPartId,
+                toProtoRpcClient(new NetworkRpcClient(this.rpcCommunicator.getRpcClientTransport()))   
+            )
+            this.targetNeighbors.add(remote)
+            this.propagation.onNeighborJoined(peer)
+            logger.info('Open proxy connection', {
+                peer
+            })
+        }
+    }
+
+    private async closeRandomConnections(connectionCount: number): Promise<void> {
+        const proxiesToDisconnect = sampleSize(Array.from(this.connections.keys()), connectionCount)
+        await Promise.allSettled(proxiesToDisconnect.map((node) => this.closeConnection(node)))
+    }
+
+    private async closeConnection(peerKey: PeerIDKey): Promise<void> {
+        if (this.connections.has(peerKey)) {
+            logger.info('Close proxy connection', {
+                peerKey
+            })
+            const server = this.targetNeighbors.getNeighborById(peerKey)
+            server?.leaveStreamNotice(this.config.ownPeerDescriptor)
+            this.removeConnection(peerKey)
+        }
+    }
+
+    private removeConnection(peerKey: PeerIDKey): void {
+        this.connections.delete(peerKey)
+        this.targetNeighbors.removeById(peerKey)
+    }
+
+    broadcast(msg: StreamMessage, previousPeer?: string): void {
+        if (!previousPeer) {
+            markAndCheckDuplicate(this.duplicateDetectors, msg.messageId!, msg.previousMessageRef)
+        }
+        this.emit('message', msg)
+        this.propagation.feedUnseenMessage(msg, this.targetNeighbors.getStringIds(), previousPeer ?? null)
+    }
+
+    getTargetNeighborStringIds(): string[] {
+        return this.targetNeighbors.getStringIds()
+    }
+
+    hasProxyConnection(peerKey: PeerIDKey, direction: ProxyDirection): boolean {
+        return this.connections.has(peerKey) && this.connections.get(peerKey) === direction
```

Comment:
The first check is redundant?

```suggestion
        return this.connections.get(peerKey) === direction
```